### DISCUSSION TITLE
Analyze issues affecting map construction

### DISCUSSION THREAD

**Discussion by HaiHai-17** [hippo5329/linorobot2_hardware#29] [General]
Hi @hippo5329 after some difficulty i got to the step of building the map but it is quite chaotic as in the picture. can you help me?
I followed every step as in the wiki and this is what I tweaked to fit the robot:
#define MOTOR_MAX_RPM 170                   // motor's max RPM
#define MAX_RPM_RATIO 0.6                  // max RPM allowed for each MAX_RPM_ALLOWED = MOTOR_MAX_RPM * MAX_RPM_RATIO
#define MOTOR_OPERATING_VOLTAGE 12          // motor's operating voltage (used to calculate max RPM)
#define MOTOR_POWER_MAX_VOLTAGE 12          // max voltage of the motor's power source (used to calculate max RPM)
#define MOTOR_POWER_MEASURED_VOLTAGE 12     // current voltage reading of the power connected to the motor (used for calibration)
#define COUNTS_PER_REV1 780                 // wheel1 encoder's no of ticks per rev
#define COUNTS_PER_REV2 780                 // wheel2 encoder's no of ticks per rev
#define COUNTS_PER_REV3 780                 // wheel3 encoder's no of ticks per rev
#define COUNTS_PER_REV4 780                 // wheel4 encoder's no of ticks per rev
#define WHEEL_DIAMETER 0.097               // wheel's diameter in meters
#define LR_WHEELS_DISTANCE 0.22           // distance between left and right wheels
#define PWM_BITS 10                         // PWM Resolution of the microcontroller
#define PWM_FREQUENCY 20000                 // PWM Frequency
#define SERVO_BITS 12                       // Servo PWM resolution
#define SERVO_FREQ 50                       // Servo PWM frequency
![Screenshot 2024-11-28 220018](https://github.com/user-attachments/assets/852461f2-129b-4c87-a877-b5f6512da625)

**Comment by HaiHai-17** [comment]
Here is what it does step by step on the wiki:
MOTOR1 FWD RPM    146.8      0.0      0.0    -48.8
MOTOR1 FWD RPM    151.1      0.0      0.0      0.0
MOTOR1 FWD RPM    151.1      0.0      0.0      0.0
MOTOR1 FWD RPM    151.2      0.0      0.0      0.0
MOTOR1 FWD RPM    151.2      0.0      0.0      0.0
MOTOR1 FWD RPM    151.2      0.0      0.0      0.0
MOTOR1 FWD RPM    151.3      0.0      0.0      0.0
MOTOR1 FWD RPM    151.2      0.0      0.0      0.0
MOTOR1 SPEED   0.77 m/s   6.98 rad/s STOP  0.290 m
MOTOR2 FWD RPM     57.1    143.2      0.0      0.0
MOTOR2 FWD RPM      0.0    147.0      0.0      0.0
MOTOR2 FWD RPM      0.0    147.1      0.0      0.0
MOTOR2 FWD RPM      0.0    147.3      0.0      0.0
MOTOR2 FWD RPM      0.0    147.3      0.0      0.0
MOTOR2 FWD RPM      0.0    147.3      0.0      0.0
MOTOR2 FWD RPM      0.0    147.2      0.0      0.0
MOTOR2 FWD RPM      0.0    147.3      0.0      0.0
MOTOR2 SPEED   0.75 m/s   6.80 rad/s STOP  0.265 m
MOTOR3 FWD RPM      0.0     52.2    145.7      0.0
MOTOR3 FWD RPM      0.0      0.0    149.2      0.0
MOTOR3 FWD RPM      0.0      0.0    149.2      0.0
MOTOR3 FWD RPM      0.0      0.0    149.2      0.0
MOTOR3 FWD RPM      0.0      0.0    149.3      0.0
MOTOR3 FWD RPM      0.0      0.0    149.5      0.0
MOTOR3 FWD RPM      0.0      0.0    149.3      0.0
MOTOR3 FWD RPM      0.0      0.0    149.3      0.0
MOTOR3 SPEED   0.76 m/s   6.89 rad/s STOP  0.255 m
MOTOR4 FWD RPM      0.0      0.0     50.2    142.5
MOTOR4 FWD RPM      0.0      0.0      0.0    146.3
MOTOR4 FWD RPM      0.0      0.0      0.0    146.5
MOTOR4 FWD RPM      0.0      0.0      0.0    146.4
MOTOR4 FWD RPM      0.0      0.0      0.0    146.5
MOTOR4 FWD RPM      0.0      0.0      0.0    146.3
MOTOR4 FWD RPM      0.0      0.0      0.0    146.6
MOTOR4 FWD RPM      0.0      0.0      0.0    146.7
MOTOR4 SPEED   0.75 m/s   6.77 rad/s STOP  0.247 m
MOTOR1 REV RPM   -141.6      0.0      0.0     48.5
MOTOR1 REV RPM   -145.2      0.0      0.0      0.0
MOTOR1 REV RPM   -145.2      0.0      0.0      0.0
MOTOR1 REV RPM   -145.3      0.0      0.0      0.0
MOTOR1 REV RPM   -145.4      0.0      0.0      0.0
MOTOR1 REV RPM   -145.4      0.0      0.0      0.0
MOTOR1 REV RPM   -145.4      0.0      0.0      0.0
MOTOR1 REV RPM   -145.3      0.0      0.0      0.0
MOTOR1 SPEED  -0.74 m/s  -6.71 rad/s STOP -0.269 m
MOTOR2 REV RPM    -53.0   -145.4      0.0      0.0
MOTOR2 REV RPM      0.0   -149.3      0.0      0.0
MOTOR2 REV RPM      0.0   -149.3      0.0      0.0
MOTOR2 REV RPM      0.0   -149.3      0.0      0.0
MOTOR2 REV RPM      0.0   -149.5      0.0      0.0
MOTOR2 REV RPM      0.0   -149.5      0.0      0.0
MOTOR2 REV RPM      0.0   -149.5      0.0      0.0
MOTOR2 REV RPM      0.0   -149.5      0.0      0.0
MOTOR2 SPEED  -0.76 m/s  -6.90 rad/s STOP -0.258 m
MOTOR3 REV RPM      0.0    -50.8   -142.2      0.0
MOTOR3 REV RPM      0.0      0.0   -145.7      0.0
MOTOR3 REV RPM      0.0      0.0   -145.8      0.0
MOTOR3 REV RPM      0.0      0.0   -145.9      0.0
MOTOR3 REV RPM      0.0      0.0   -145.8      0.0
MOTOR3 REV RPM      0.0      0.0   -145.8      0.0
MOTOR3 REV RPM      0.0      0.0   -145.8      0.0
MOTOR3 REV RPM      0.0      0.0   -145.8      0.0
MOTOR3 SPEED  -0.74 m/s  -6.73 rad/s STOP -0.253 m
MOTOR4 REV RPM      0.0      0.0    -49.8   -145.7
MOTOR4 REV RPM      0.0      0.0      0.0   -149.3
MOTOR4 REV RPM      0.0      0.0      0.0   -149.5
MOTOR4 REV RPM      0.0      0.0      0.0   -149.5
MOTOR4 REV RPM      0.0      0.0      0.0   -149.5
MOTOR4 REV RPM      0.0      0.0      0.0   -149.7
MOTOR4 REV RPM      0.0      0.0      0.0   -149.6
MOTOR4 REV RPM      0.0      0.0      0.0   -149.6
MOTOR4 SPEED  -0.76 m/s  -6.91 rad/s STOP -0.249 m


guen@raspberry:~/guenbot_hardware/firmware$ ros2 topic hz /imu/data
average rate: 50.127
        min: 0.017s max: 0.022s std dev: 0.00088s window: 52
average rate: 50.102
        min: 0.017s max: 0.028s std dev: 0.00129s window: 103
average rate: 50.066
        min: 0.017s max: 0.028s std dev: 0.00111s window: 154

guen@raspberry:~/guenbot_hardware/firmware$ ros2 topic hz /odom
average rate: 50.143
        min: 0.018s max: 0.022s std dev: 0.00059s window: 52
average rate: 50.069
        min: 0.018s max: 0.022s std dev: 0.00055s window: 102
average rate: 50.046
        min: 0.018s max: 0.022s std dev: 0.00057s window: 153

guen@raspberry:~/guenbot_hardware/firmware$ ros2 topic hz /joint_states 
average rate: 10.017
        min: 0.098s max: 0.101s std dev: 0.00066s window: 12
average rate: 10.010
        min: 0.098s max: 0.101s std dev: 0.00050s window: 23
average rate: 10.006
        min: 0.098s max: 0.101s std dev: 0.00042s window: 34

guen@raspberry:~/guenbot_hardware/firmware$ ros2 topic hz /scan 
average rate: 12.038
        min: 0.073s max: 0.091s std dev: 0.00623s window: 14
average rate: 12.067
        min: 0.073s max: 0.091s std dev: 0.00573s window: 27
average rate: 12.079
        min: 0.073s max: 0.091s std dev: 0.00551s window: 40

**Comment by hippo5329** [comment]
max rpm should be 150 for your motor.

    #define MOTOR_MAX_RPM 150                   // motor's max RPM
    #define MAX_RPM_RATIO 0.85                  // max RPM allowed for each MAX_RPM_ALLOWED = MOTOR_MAX_RPM * MAX_RPM_RATIO

stop distance is too long, 0.25m = 25cm.  which motor and motor driver do you use?

    #define USE_SHORT_BRAKE // for shorter stopping distance

**Reply by hippo5329** [reply]
run test_acc in open space.

**Reply by HaiHai-17** [reply]
i am using ga25-370 170rpm and l298n, i think the long stopping distance is because of the l298 module, because the room area is a bit small but i will try test_acc
https://dientutuyetnga.com/products/ga25-370-encoder-dong-co-giam-toc-170rpm

**Reply by hippo5329** [reply]
test_sensors

**Reply by HaiHai-17** [reply]
it's fine and i pulled 2k for scl and sda
ACC 0.12 0.00 9.83 GYR 0.00 0.00 0.00
ACC 0.01 0.02 9.78 GYR 0.00 0.00 0.00
ACC -0.03 -0.01 9.81 GYR 0.00 0.00 0.00
![image](https://github.com/user-attachments/assets/70952aa7-400c-41fa-86fb-8d2018e70ff5)

**Comment by hippo5329** [comment]
Your motor is 150rpm variant.

L298 does support fast motor stop.

For Mecanum drive, the LR_WHEELS_DISTANCE should somewhere as front-right to rear-left.

**Reply by HaiHai-17** [reply]
I set it to 150rpm and LR_WHEELS_DISTANCE 0.22

**Comment by hippo5329** [comment]
How much is your robot weight?

**Reply by HaiHai-17** [reply]
It weighs 3.1 kg, is 25 long, 15 wide, and 20 high.

**Reply by HaiHai-17** [reply]
If including wheel width then 25x27x20

**Comment by HaiHai-17** [comment]
When launching bringup.launch.py ​​it shows the following error message:
[ekf_node-1] Failed to meet update rate! Took 0.023976464999999998889seconds. Try decreasing the rate, limiting sensor output frequency, or limiting the number of sensors.
[joint_state_publisher-4] [INFO] [1732815545.570153017] [joint_state_publisher]: Waiting for robot_description to be published on the robot_description topic...
[ekf_node-1] Failed to meet update rate! Took 0.054008035000000002845seconds. Try decreasing the rate, limiting sensor output frequency, or limiting the number of sensors.

**Comment by hippo5329** [comment]
You should check the real rpm with a tachometer if you can find one. Or you may try tachometer/strobe light app on your phone.

22 is too small for your Mecanum driver. What is the distance of front-right wheel to front-left? front-right to rear-right? Then you can find the trigonometry.

When you do the slam, do not turn, only forward and backward. You can see the error between laser and your map. This is the linear wheel odometry error. Next, turn only. You can see the rotation error.

**Reply by HaiHai-17** [reply]
this is when i show forward and backward
![Screenshot 2024-11-29 012141](https://github.com/user-attachments/assets/50440567-ce8a-4a9b-8b37-e88aca72e2d5)
this is when i rotated in place and noticed the problem as you mentioned
![Screenshot 2024-11-29 012212](https://github.com/user-attachments/assets/577cf153-3432-4acb-943c-6f68ea058939)

**Reply by HaiHai-17** [reply]
so now i need to re-measure the distance between the wheels and readjust the rpm, actual speed

**Comment by HaiHai-17** [comment]
Hi @hippo5329 today my motors 1-3 are spinning continuously even though they are not receiving control commands, motors 2-4 do not have the same problem when running the firmware. test_motor shows that motors 1-2-3-4 are still working normally and functioning properly.

**Comment by hippo5329** [comment]
During developments process, things went wrong often. It will be helpful to revert to previous working state. You can use git to keep track your code changes. You can revert your changes with commits or try different alternatives with git branches.

I use both KVM snapshots and LXD/ZFS snapshots to create snapshots at each check points. So that I can revert to previous working state easily. I use wired connection and create a bridge to connect them. Because ROS2 RMW uses broadcast. So bridged connections works best.

In your case, you should save current work and revert to previous working state.

**Reply by HaiHai-17** [reply]
i know where my cause is, because i switched to pico and tb6612. when plugging in the encoder pins of motor 1-3 must be reversed with 2-4. 
because motor 1-3 is flipped to the opposite side, this is basic knowledge that i forgot, i think should have more notes for myself and others. 
thank you very much for supporting me.

**Comment by hippo5329** [comment]
You should run test_motors after you modify the motors/encoders connection.

**Reply by HaiHai-17** [reply]
@hippo5329  map creation is working pretty well, thanks for your support.
Now I'm starting with the camera but it seems to be working pretty slow, I'm using ov5647 camera and have 1 usb webcam.
I see in your model there is a usb webcam, can you help me, which one and which package should I use?
![Screenshot 2024-12-06 233304](https://github.com/user-attachments/assets/04197091-1f10-443d-af2c-97a289a483c4)

**Reply by hippo5329** [reply]
I had troubles with libcamera support of  pi camera on Ubuntu before. So I used usb camera. There are less issues with usb camera drivers. What do you mean "slow"? I will try ov5647 again later.

**Reply by HaiHai-17** [reply]
I'm also having problems with libcamera so when I switched to v4l2 the camera's fps was only about 3-4fps. By "slow" I mean fps and sometimes it even freezes.

**Comment by hippo5329** [comment]
Follow this link to build libcamera,

https://www.raspberrypi.com/documentation/computers/camera_software.html#building-libcamera

You will need to install missing dependencies.

I am still working on this.

**Reply by hippo5329** [reply]
```
ubuntu@raspi5:~/rpicam-apps$ sudo rpicam-hello
[0:35:12.242297775] [8548]  INFO Camera camera_manager.cpp:325 libcamera v0.3.2+99-1230f78d
[0:35:12.262965834] [8551]  INFO RPI pisp.cpp:695 libpisp version v1.0.7 28196ed6edcf 07-12-2024 (13:07:52)
[0:35:12.277994142] [8551]  INFO RPI pisp.cpp:1154 Registered camera /base/axi/pcie@120000/rp1/i2c@88000/ov5647@36 to CFE device /dev/media0 and ISP device /dev/media2 using PiSP variant BCM2712_C0
Preview window unavailable
[0:35:12.278655167] [8548]  WARN V4L2 v4l2_pixelformat.cpp:346 Unsupported V4L2 pixel format RPBP
Mode selection for 1296:972:12:P
    SGBRG10_CSI2P,640x480/0 - Score: 3296
    SGBRG10_CSI2P,1296x972/0 - Score: 1000
    SGBRG10_CSI2P,1920x1080/0 - Score: 1349.67
    SGBRG10_CSI2P,2592x1944/0 - Score: 1567
Stream configuration adjusted
[0:35:12.278968429] [8548]  INFO Camera camera.cpp:1197 configuring streams: (0) 1296x972-YUV420 (1) 1296x972-GBRG_PISP_COMP1
[0:35:12.279097405] [8551]  INFO RPI pisp.cpp:1450 Sensor: /base/axi/pcie@120000/rp1/i2c@88000/ov5647@36 - Selected sensor format: 1296x972-SGBRG10_1X10 - Selected CFE format: 1296x972-PC1g
#0 (0.00 fps) exp 29995.00 ag 1.62 dg 1.06
#1 (32.42 fps) exp 29995.00 ag 1.62 dg 1.06
#2 (32.42 fps) exp 29995.00 ag 1.62 dg 1.06
#3 (32.42 fps) exp 29995.00 ag 1.69 dg 1.02
#4 (32.42 fps) exp 29995.00 ag 1.69 dg 1.02
#5 (32.42 fps) exp 29995.00 ag 1.69 dg 1.01
#6 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#7 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#8 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#9 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#10 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#11 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#12 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#13 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#14 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#15 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#16 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#17 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#18 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#19 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#20 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#21 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#22 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#23 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#24 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#25 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#26 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#27 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#28 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#29 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#30 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#31 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#32 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#33 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#34 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#35 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#36 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#37 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#38 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#39 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#40 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#41 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#42 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#43 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#44 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#45 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#46 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#47 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#48 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#49 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#50 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#51 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#52 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#53 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#54 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#55 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#56 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#57 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#58 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#59 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#60 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#61 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#62 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#63 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#64 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#65 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#66 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#67 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#68 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#69 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#70 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#71 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#72 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#73 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#74 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#75 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#76 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#77 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#78 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#79 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#80 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#81 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#82 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#83 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#84 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#85 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#86 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#87 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#88 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#89 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#90 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#91 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#92 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#93 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#94 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#95 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#96 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#97 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#98 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#99 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#100 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#101 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#102 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#103 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#104 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#105 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#106 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#107 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#108 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#109 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#110 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#111 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#112 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#113 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#114 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#115 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#116 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#117 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#118 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#119 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#120 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#121 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#122 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#123 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#124 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#125 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#126 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#127 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#128 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#129 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#130 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#131 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#132 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#133 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#134 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#135 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#136 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#137 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#138 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#139 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#140 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#141 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#142 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#143 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#144 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#145 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#146 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#147 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#148 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#149 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#150 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#151 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#152 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
#153 (32.42 fps) exp 29995.00 ag 1.75 dg 1.00
ubuntu@raspi5:~/rpicam-apps$ 


```

**Reply by hippo5329** [reply]
I think it is mostly working now.

**Reply by HaiHai-17** [reply]
it working, I am looking for source and testing developing following human robot on ros2 using rgb camera
[0:07:22.341140931] [2780]  INFO Camera camera_manager.cpp:325 libcamera v0.3.2+99-1230f78d
[0:07:22.395345783] [2783]  WARN RPiSdn sdn.cpp:40 Using legacy SDN tuning - please consider moving SDN inside rpi.denoise
[0:07:22.400374023] [2783]  INFO RPI vc4.cpp:447 Registered camera /base/soc/i2c0mux/i2c@1/ov5647@36 to Unicam device /dev/media0 and ISP device /dev/media1
[0:07:22.400496093] [2783]  INFO RPI pipeline_base.cpp:1120 Using configuration file '/usr/local/share/libcamera/pipeline/rpi/vc4/rpi_apps.yaml'
Preview window unavailable
Mode selection for 1296:972:12:P
    SGBRG10_CSI2P,640x480/0 - Score: 3296
    SGBRG10_CSI2P,1296x972/0 - Score: 1000
    SGBRG10_CSI2P,1920x1080/0 - Score: 1349.67
    SGBRG10_CSI2P,2592x1944/0 - Score: 1567
Stream configuration adjusted
[0:07:22.403129004] [2780]  INFO Camera camera.cpp:1197 configuring streams: (0) 1296x972-YUV420 (1) 1296x972-SGBRG10_CSI2P
[0:07:22.405111016] [2783]  INFO RPI vc4.cpp:622 Sensor: /base/soc/i2c0mux/i2c@1/ov5647@36 - Selected sensor format: 1296x972-SGBRG10_1X10 - Selected unicam format: 1296x972-pGAA
#0 (0.00 fps) exp 33222.00 ag 8.00 dg 1.00
#1 (30.02 fps) exp 33222.00 ag 8.00 dg 1.00
#2 (30.02 fps) exp 33222.00 ag 8.00 dg 1.00
#3 (30.02 fps) exp 33222.00 ag 8.00 dg 1.00
#4 (30.02 fps) exp 33222.00 ag 8.00 dg 1.00
#5 (30.02 fps) exp 33222.00 ag 8.00 dg 1.00
#6 (30.02 fps) exp 33222.00 ag 8.00 dg 1.00
#7 (30.02 fps) exp 33222.00 ag 8.00 dg 1.00
#8 (30.01 fps) exp 33222.00 ag 8.00 dg 1.00
#9 (30.03 fps) exp 33222.00 ag 8.00 dg 1.00
#10 (30.01 fps) exp 33222.00 ag 8.00 dg 1.00

**Comment by johnny-0216** [comment]


**Reply by hippo5329** [reply]
Please create a new discussion and describe your configuration in more detail. Every case might be different.

**Reply by johnny-0216** [reply]
Ok
